File: code/BenchmarkCpp/StrassenBenchmark.cpp
#include "MatrixAlgorithm.h"
#include <iostream>
#include <vector>
#include <chrono>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <ctime>
#include <cstdlib>

using namespace std;

// ===================================================
// PHẦN 1: TEST CASES
// ===================================================

void runTestCases() {
    /**
     * Chạy bộ kiểm thử đầy đủ và in kết quả
     * Kiểm tra tính đúng đắn của Strassen so với Naive
     */
    cout << "\n" << string(60, '=') << "\n";
    cout << "CHAY CAC TEST CASES\n";
    cout << string(60, '=') << "\n";
    
    int passed = 0;
    int total = 0;
    
    // Test 1: Ma trận nhỏ 2x2
    {
        total++;
        Matrix A = {{1, 2}, {3, 4}};
        Matrix B = {{5, 6}, {7, 8}};
        Matrix expected = {{19, 22}, {43, 50}};
        Matrix result = strassenGeneral(A, B);
        double err = matrixMaxAbsError(result, expected);
        
        cout << "\nTest 1: Ma tran 2x2\n";
        cout << "  Ket qua: " << (err < 1e-9 ? "✓ PASS" : "✗ FAIL");
        cout << " (Max error = " << scientific << err << ")\n";
        if (err < 1e-9) passed++;
    }
    
    // Test 2: Ma trận ngẫu nhiên 64x64
    {
        total++;
        Matrix A = generateRandomMatrix(64, 64);
        Matrix B = generateRandomMatrix(64, 64);
        Matrix result_naive = matrixMultiplyNaive(A, B);
        Matrix result_strassen = strassenGeneral(A, B);
        double err = matrixMaxAbsError(result_naive, result_strassen);
        
        cout << "\nTest 2: Ma tran ngau nhien 64x64\n";
        cout << "  Ket qua: " << (err < 1e-6 ? "✓ PASS" : "✗ FAIL");
        cout << " (Max error = " << scientific << err << ")\n";
        if (err < 1e-6) passed++;
    }
    
    // Test 3: Ma trận toàn số 0
    {
        total++;
        Matrix A(4, vector<double>(4, 0.0));
        Matrix B(4, vector<double>(4, 0.0));
        Matrix result = strassenGeneral(A, B);
        
        bool all_zero = true;
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                if (abs(result[i][j]) > 1e-12) {
                    all_zero = false;
                    break;
                }
            }
        }
        
        cout << "\nTest 3: Ma tran toan so 0 (4x4)\n";
        cout << "  Ket qua: " << (all_zero ? "✓ PASS" : "✗ FAIL");
        cout << " (Tat ca phan tu = 0)\n";
        if (all_zero) passed++;
    }
    
    // Test 4: Ma trận đơn vị (Identity)
    {
        total++;
        int n = 4;
        Matrix I(n, vector<double>(n, 0.0));
        for (int i = 0; i < n; i++) I[i][i] = 1.0;
        
        Matrix A = generateRandomMatrix(n, n);
        Matrix result = strassenGeneral(A, I);
        double err = matrixMaxAbsError(A, result);
        
        cout << "\nTest 4: Ma tran đon vi 4x4 (A*I=A)\n";
        cout << "  Ket qua: " << (err < 1e-9 ? "✓ PASS" : "✗ FAIL");
        cout << " (Max error = " << scientific << err << ")\n";
        if (err < 1e-9) passed++;
    }
    
    // Test 5: Ma trận 1x1
    {
        total++;
        Matrix A = {{7.0}};
        Matrix B = {{3.0}};
        Matrix result = strassenGeneral(A, B);
        double err = abs(result[0][0] - 21.0);
        
        cout << "\nTest 5: Ma tran 1x1 (7*3=21)\n";
        cout << "  Ket qua: " << (err < 1e-9 ? "✓ PASS" : "✗ FAIL");
        cout << " (Ket qua = " << fixed << setprecision(1) << result[0][0] << ")\n";
        if (err < 1e-9) passed++;
    }
    
    // Test 6: Ma trận kích thước lẻ 5x5
    {
        total++;
        Matrix A = generateRandomMatrix(5, 5);
        Matrix B = generateRandomMatrix(5, 5);
        Matrix result_naive = matrixMultiplyNaive(A, B);
        Matrix result_strassen = strassenGeneral(A, B);
        double err = matrixMaxAbsError(result_naive, result_strassen);
        
        cout << "\nTest 6: Ma tran kich thuoc le 5x5\n";
        cout << "  Ket qua: " << (err < 1e-6 ? "✓ PASS" : "✗ FAIL");
        cout << " (Max error = " << scientific << err << ")\n";
        if (err < 1e-6) passed++;
    }
    
    // Test 7: Ma trận không vuông (3x5 × 5x4)
    {
        total++;
        Matrix A = generateRandomMatrix(3, 5);
        Matrix B = generateRandomMatrix(5, 4);
        Matrix result_naive = matrixMultiplyNaive(A, B);
        Matrix result_strassen = strassenGeneral(A, B);
        double err = matrixMaxAbsError(result_naive, result_strassen);
        
        cout << "\nTest 7: Ma tran khong vuong 3×5 × 5×4\n";
        cout << "  Ket qua: " << (err < 1e-6 ? "✓ PASS" : "✗ FAIL");
        cout << " (Max error = " << scientific << err << ")\n";
        if (err < 1e-6) passed++;
    }
    
    cout << "\n" << string(60, '-') << "\n";
    cout << "Tong: " << passed << "/" << total << " test cases PASSED\n\n";
}

// ===================================================
// PHẦN 2: BENCHMARK
// ===================================================

struct BenchmarkResult {
    int size;
    double time_naive;
    double time_strassen;
    double time_strassen_hybrid;
};

vector<BenchmarkResult> benchmarkAlgorithms(vector<int> sizes, int repeats) {
    /**
     * So sánh thời gian chạy của các thuật toán:
     *   1. Naive (vòng lặp C++)
     *   2. Strassen (đệ quy thuần túy)
     *   3. Strassen Hybrid
     */
    vector<BenchmarkResult> results;
    
    cout << "\n" << string(110, '=') << "\n";
    cout << "BENCHMARK: SO SANH CAC THUAT TOAN NHAN MA TRAN\n";
    cout << string(110, '=') << "\n";
    cout << fixed << setprecision(4);
    cout << setw(8) << "n" 
         << setw(16) << "Naive (s)"
         << setw(16) << "Strassen (s)"
         << setw(20) << "Strassen Hybrid (s)"
         << setw(14) << "Speedup S/N"
         << setw(14) << "Speedup SH/N"
         << "\n";
    cout << string(110, '-') << "\n";
    
    for (int n : sizes) {
        // Sinh ma trận ngẫu nhiên
        Matrix A = generateRandomMatrix(n, n);
        Matrix B = generateRandomMatrix(n, n);
        
        // Đo thời gian Naive
        auto measure_naive = [&]() { matrixMultiplyNaive(A, B); };
        double t_naive = 0;
        for (int i = 0; i < repeats; i++) {
            auto start = chrono::high_resolution_clock::now();
            measure_naive();
            auto end = chrono::high_resolution_clock::now();
            chrono::duration<double> elapsed = end - start;
            t_naive += elapsed.count();
        }
        t_naive /= repeats;
        
        // Đo thời gian Strassen
        auto measure_strassen = [&]() { strassen(A, B); };
        double t_strassen = 0;
        for (int i = 0; i < repeats; i++) {
            auto start = chrono::high_resolution_clock::now();
            measure_strassen();
            auto end = chrono::high_resolution_clock::now();
            chrono::duration<double> elapsed = end - start;
            t_strassen += elapsed.count();
        }
        t_strassen /= repeats;
        
        // Đo thời gian Strassen Hybrid
        auto measure_hybrid = [&]() { strassenHybrid(A, B, 64); };
        double t_hybrid = 0;
        for (int i = 0; i < repeats; i++) {
            auto start = chrono::high_resolution_clock::now();
            measure_hybrid();
            auto end = chrono::high_resolution_clock::now();
            chrono::duration<double> elapsed = end - start;
            t_hybrid += elapsed.count();
        }
        t_hybrid /= repeats;
        
        // Tính speedup
        double speedup_s = (t_naive > 0) ? (t_naive / t_strassen) : 0;
        double speedup_h = (t_naive > 0) ? (t_naive / t_hybrid) : 0;
        
        cout << setw(8) << n
             << setw(16) << t_naive
             << setw(16) << t_strassen
             << setw(20) << t_hybrid
             << setw(14) << (speedup_s > 0 ? to_string(speedup_s) + "x" : "N/A")
             << setw(14) << (speedup_h > 0 ? to_string(speedup_h) + "x" : "N/A")
             << "\n";
        
        results.push_back({n, t_naive, t_strassen, t_hybrid});
    }
    
    cout << string(110, '=') << "\n\n";
    return results;
}

void exportResultsToCSV(const vector<BenchmarkResult>& results, const string& filename) {
    /**
     * Xuất kết quả benchmark ra file CSV
     */
    ofstream file(filename);
    file << "Size,Naive_Time,Strassen_Time,Strassen_Hybrid_Time\n";
    
    for (const auto& r : results) {
        file << r.size << ","
             << fixed << setprecision(6)
             << r.time_naive << ","
             << r.time_strassen << ","
             << r.time_strassen_hybrid << "\n";
    }
    
    file.close();
    cout << "[✓] Ket qua CSV da luu: " << filename << "\n";
}

void exportResultsToJSON(const vector<BenchmarkResult>& results, const string& filename) {
    /**
     * Xuất kết quả benchmark ra file JSON
     */
    ofstream file(filename);
    file << "[\n";
    
    for (size_t i = 0; i < results.size(); i++) {
        const auto& r = results[i];
        file << "  {\n"
             << "    \"size\": " << r.size << ",\n"
             << "    \"naive_time\": " << fixed << setprecision(6) << r.time_naive << ",\n"
             << "    \"strassen_time\": " << r.time_strassen << ",\n"
             << "    \"strassen_hybrid_time\": " << r.time_strassen_hybrid << "\n"
             << "  }";
        if (i < results.size() - 1) file << ",";
        file << "\n";
    }
    
    file << "]\n";
    file.close();
    cout << "[✓] Ket qua JSON da luu: " << filename << "\n";
}

// ===================================================
// MAIN — CHẠY TOÀN BỘ
// ===================================================

int main() {
    // Khởi tạo random seed
    srand(time(0));
    
    cout << "\n" << string(70, '=') << "\n";
    cout << "NHAN MA TRAN STRASSEN — BAO CAO THUC NGHIEM (C++)\n";
    cout << string(70, '=') << "\n";
    
    // 1. Test cases
    runTestCases();
    
    // 2. Benchmark
    vector<int> sizes = {32, 64, 128, 256, 512};
    int repeats = 3;
    cout << "Kich thuoc thu: ";
    for (int s : sizes) cout << s << " ";
    cout << "\nSo lan lap moi kich thuoc: " << repeats << "\n\n";
    
    vector<BenchmarkResult> results = benchmarkAlgorithms(sizes, repeats);
    
    // 3. Xuất kết quả
    string result_dir = "D:\\DAA\\Project\\result\\BenchmarkCpp\\";
    exportResultsToCSV(results, result_dir + "benchmark_results_cpp.csv");
    exportResultsToJSON(results, result_dir + "benchmark_results_cpp.json");
    
    cout << "\nHoan thanh. Ket qua da luu.\n";
    
    return 0;
}

File: code/BenchmarkJava/StrassenBenchmark.java
package BenchmarkJava;
import java.util.*;

/**
 * Main class - Chạy toàn bộ benchmark và test cases
 */
public class StrassenBenchmark {
    
    /**
     * Chạy các test cases để kiểm tra tính đúng đắn
     */
    public static void runTestCases() {
        System.out.println("=" .repeat(20));
        System.out.println("CHAY CAC TEST CASES");
        System.out.println("=" .repeat(20));
        
        List<Object[]> results = new ArrayList<>();
        
        try {
            // Test 1: Ma trận 2x2
            double[][] A1 = {{1, 2}, {3, 4}};
            double[][] B1 = {{5, 6}, {7, 8}};
            double[][] C1_naive = MatrixAlgorithm.matrixMultiplyNaive(A1, B1);
            double[][] C1_strassen = MatrixAlgorithm.strassenGeneral(A1, B1, 64);
            double err1 = MatrixOperation.matrixMaxAbsError(C1_naive, C1_strassen);
            results.add(new Object[]{"Ma tran 2x2", err1 < 1e-9, String.format("Max error = %.2e", err1)});
            
            // Test 2: Ma trận ngẫu nhiên 64x64
            double[][] A2 = MatrixOperation.generateRandomMatrix(64, 64, -10, 10);
            double[][] B2 = MatrixOperation.generateRandomMatrix(64, 64, -10, 10);
            double[][] C2_naive = MatrixAlgorithm.matrixMultiplyNaive(A2, B2);
            double[][] C2_strassen = MatrixAlgorithm.strassenGeneral(A2, B2, 64);
            double err2 = MatrixOperation.matrixMaxAbsError(C2_naive, C2_strassen);
            results.add(new Object[]{"Ma tran ngau nhien 64x64", err2 < 1e-6, String.format("Max error = %.2e", err2)});
            
            // Test 3: Ma trận toàn số 0
            double[][] A3 = new double[4][4];
            double[][] B3 = new double[4][4];
            double[][] C3 = MatrixAlgorithm.strassenGeneral(A3, B3, 64);
            boolean isZero = true;
            for (int i = 0; i < 4; i++) {
                for (int j = 0; j < 4; j++) {
                    if (Math.abs(C3[i][j]) > 1e-12) {
                        isZero = false;
                    }
                }
            }
            results.add(new Object[]{"Ma tran toan so 0 (4x4)", isZero, "Tat ca phan tu = 0"});
            
            // Test 4: Ma trận đơn vị
            double[][] A4 = MatrixOperation.generateRandomMatrix(4, 4, -10, 10);
            double[][] I = new double[4][4];
            for (int i = 0; i < 4; i++) {
                I[i][i] = 1.0;
            }
            double[][] C4 = MatrixAlgorithm.strassenGeneral(A4, I, 64);
            double err4 = MatrixOperation.matrixMaxAbsError(C4, A4);
            results.add(new Object[]{"Ma tran đon vi 4x4 (A*I=A)", err4 < 1e-9, String.format("Max error = %.2e", err4)});
            
            // Test 5: Ma trận 1x1
            double[][] A5 = {{7.0}};
            double[][] B5 = {{3.0}};
            double[][] C5 = MatrixAlgorithm.strassenGeneral(A5, B5, 64);
            double err5 = Math.abs(C5[0][0] - 21.0);
            results.add(new Object[]{"Ma tran 1x1 (7*3=21)", err5 < 1e-9, String.format("Ket qua = %.1f", C5[0][0])});
            
            // Test 6: Ma trận kích thước lẻ 5x5
            double[][] A6 = MatrixOperation.generateRandomMatrix(5, 5, -10, 10);
            double[][] B6 = MatrixOperation.generateRandomMatrix(5, 5, -10, 10);
            double[][] C6_naive = MatrixAlgorithm.matrixMultiplyNaive(A6, B6);
            double[][] C6_strassen = MatrixAlgorithm.strassenGeneral(A6, B6, 64);
            double err6 = MatrixOperation.matrixMaxAbsError(C6_naive, C6_strassen);
            results.add(new Object[]{"Ma tran kich thuoc le 5x5", err6 < 1e-6, String.format("Max error = %.2e", err6)});
            
            // Test 7: Ma trận không vuông 3x5 × 5x4
            double[][] A7 = MatrixOperation.generateRandomMatrix(3, 5, -10, 10);
            double[][] B7 = MatrixOperation.generateRandomMatrix(5, 4, -10, 10);
            double[][] C7_naive = MatrixAlgorithm.matrixMultiplyNaive(A7, B7);
            double[][] C7_strassen = MatrixAlgorithm.strassenGeneral(A7, B7, 64);
            double err7 = MatrixOperation.matrixMaxAbsError(C7_naive, C7_strassen);
            results.add(new Object[]{"Ma tran khong vuong 3×5 × 5×4", err7 < 1e-6, String.format("Max error = %.2e", err7)});
            
        } catch (Exception e) {
            System.err.println("Loi trong test case: " + e.getMessage());
            e.printStackTrace();
        }
        
        // In kết quả
        System.out.printf("\n%-40s | %-10s | %s\n", "Test case", "Ket qua", "Chi tiet");
        System.out.println("-" .repeat(70));
        
        int passed = 0;
        for (Object[] result : results) {
            String name = (String) result[0];
            boolean ok = (Boolean) result[1];
            String detail = (String) result[2];
            String status = ok ? "✓ PASS" : "✗ FAIL";
            
            System.out.printf("%-40s | %-10s | %s\n", name, status, detail);
            if (ok) passed++;
        }
        
        System.out.println("-" .repeat(70));
        System.out.printf("Tong: %d/%d test cases PASSED\n\n", passed, results.size());
    }
    
    /**
     * Main function
     */
    public static void main(String[] args) {
        System.out.println("\n" + "=" .repeat(70));
        System.out.println("  NHAN MA TRAN STRASSEN — BAO CAO THUC NGHIEM");
        System.out.println("=" .repeat(70));
        
        // 1. Test cases
        runTestCases();
        
        // 2. Benchmark
        int[] SIZES = {32, 64, 128, 256, 512, 1024};
        int REPEATS = 10;
        
        System.out.println("Kich thuoc thu: " + Arrays.toString(SIZES));
        System.out.println("So lan lap moi kich thuoc: " + REPEATS + "\n");
        
        List<BenchmarkResult> results = Benchmark.runBenchmark(SIZES, REPEATS);
        
        // 3. In bảng kết quả chi tiết
        System.out.println("\n--- BANG KET QUA CHI TIET ---");
        System.out.printf("%-12s | %-15s | %-15s | %-20s | %-20s | %-20s\n",
            "Kich thuoc", "Naive (s)", "Strassen (s)", "Strassen Hybrid (s)", 
            "Speedup S/N", "Speedup SH/N");
        System.out.println("-" .repeat(110));
        
        for (BenchmarkResult result : results) {
            System.out.printf("%-12d | %-15.6f | %-15.6f | %-20.6f | %-20.2f | %-20.2f\n",
                result.getSize(),
                result.getNaiveTime(),
                result.getStrassenTime(),
                result.getStrassenHybridTime(),
                result.getSpeedupStrassen(),
                result.getSpeedupStrassenHybrid()
            );
        }
        System.out.println("-" .repeat(110));
        
        // 4. Xuất kết quả ra CSV và JSON
        String resultDir = "d:\\DAA\\Project\\result\\BenchmarkJava\\";
        ResultExporter.exportToCSV(results, resultDir + "benchmark_results.csv");
        ResultExporter.exportToJSON(results, resultDir + "benchmark_results.json");
        
        System.out.println("\nHoan thanh. Ket qua da luu.");
    }
}
